### ISSUE TITLE
Add hello-world and todo-mvc Quarkus examples

### ISSUE THREAD

**Issue by lihaoyi** [com-lihaoyi/mill#3549]
Mill's example test suite has examples for hello-world and todo-mvc projects using in Java using Spring-Boot and Micronaut built using Mill (https://github.com/com-lihaoyi/mill/tree/main/example/javalib/web). They're not super fleshed out, but they exist.

We should add similar hello-world/todo-mvc examples for the Quarkus web framework, which according to the IntelliJ dev survey is relatively popular (between Spring-Boot and Micronaut https://www.jetbrains.com/lp/devecosystem-2023/java/)

**Comment by lolgab** [comment]
I'm giving a shot to this.

**Comment by lefou** [comment]
The `import`-scope Maven dependencies are called BOM - Bill of Material. There are open requests for it in Mill and coursier:

* https://github.com/com-lihaoyi/mill/issues/1975
* https://github.com/coursier/coursier/issues/1390

I think @alexarchambault will work on it in the next months (https://github.com/com-lihaoyi/mill/discussions/3496).

**Comment by llvee** [comment]
@lolgab Did you complete this? If not maybe I will give this a shot in 24 hours from now.

**Comment by llvee** [comment]
@lolgab Would you like some help with it? Have you made a lot of progress?

**Comment by omarsoufiane** [comment]
hello @lihaoyi,
I'm the founder of [Bountyhub](https://bountyhub.dev), a new platform dedicated exclusively to managing bounties on GitHub issues. Our platform is designed to provide greater transparency between bounty creators and solvers, ensuring that both sides have clear visibility throughout the process. people who use your projects can also create bounties on your repositories. 

we would be very happy if you install our github app and add this badge to your readme:
[![Bounties on BountyHub](https://img.shields.io/badge/Bounties-on%20BountyHub-yellow)]( https://bountyhub.dev/bounties?repo=com-lihaoyi/mill)
it will redirect contributors to the bounties created for this repository (try it by clicking on the badge above).

Looking forward to hearing your thoughts.

**Comment by lolgab** [comment]
No, I haven't. Feel free to work on it 👍

**Comment by lihaoyi** [comment]
Bumping this up from 500 to 1000USD

**Comment by himanshumahajan138** [comment]
@lihaoyi Working on this (Have Read Docs and Seems it can be done)

**Comment by himanshumahajan138** [comment]
> quarkus, in its maven and gradle plugins, does some code generation which seems to be necessary to create the `main` method to run applications.

@lolgab could u please provide me some documentation or some link which will guide me for this 

Actually i got some understanding but not properly able to understand what's going on internally 
please give some brief if possible

**Comment by himanshumahajan138** [comment]
@lolgab means we have to create proper plugin first of all then use that plugin as dependency in our examples

what u think we can use the already available codes in the link you gave 

looks hard than it looks 😶

**Comment by himanshumahajan138** [comment]
@lolgab what if we use the **"quarkus cli"** for creating and running the app

**Comment by lolgab** [comment]
It could work but it's very inefficient since the cli  is a thin wrapper to the Maven plugin. So mill will call Maven under the hood which is not ideal.
Yes, it's harder than it looks, but it's totally doable if you have the motivation.
At first you can implement the plugin as part of the example build, extracting as a contrib module is easy afterwards.

**Comment by himanshumahajan138** [comment]
@lolgab I know i should not ask like this but i am motivated and have will to complete this so i need some help regarding this could u please provide me some outer guide like roadmap to reach to end point not only for bounty but in a sense to learn something professionally 

TBH I can do this but don't know how to start and what reference to take as example to work on this 

Hope U Understand !!!

**Comment by JWLZGG** [comment]
@lihaoyi is this bounty this active?

**Comment by Lynndabel** [comment]
@lihaoyi can i work on this?

**Comment by lihaoyi** [comment]
Removing the bounty since after over a year we haven't managed to merge it

**Comment by vaslabs** [comment]
The main challenge for Quarkus is its integration points for generating [the application models](https://github.com/quarkusio/quarkus/blob/main/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/ApplicationModelSerializer.java) . Quarkus seems to only offer out of the box support for gradle and maven (they publish the plugins themselves) and we can see in the model serializer class things like

```java
    public static void exportGradleModel(ApplicationModel model, boolean test) throws IOException {
        System.setProperty(test ? BootstrapConstants.SERIALIZED_TEST_APP_MODEL : BootstrapConstants.SERIALIZED_APP_MODEL,
                serializeGradleModel(model, test).toString());
    }
```

My main idea, very abstractly is that we need to create a mill worker integration that creates the ApplicationModel and serializes it with 

```java
    public static void serialize(ApplicationModel appModel, Path file) throws IOException {
        Files.createDirectories(file.getParent());
        if (JOS) {
            serializeWithJos(appModel, file);
        } else {
            toJson(appModel, file);
        }
    }
```

I have not found any other integration points, the CLI itself uses either gradle or maven so it's not something we can reuse.


Then, after we make Quarkus generate the serialized application model files, we can use all the utilities (such as running the integration tests)

**Comment by lefou** [comment]
It's always the same pattern, frameworks come with some selected tool integrations, while they should just provide some standalone tooling that can be easily integrated in tools afterwards, either as CLI app or as a plugin. I'm so sick of seeing it again and again.

**Comment by vaslabs** [comment]
> It's always the same pattern, frameworks come with some selected tool integrations, while they should just provide some standalone tooling that can be easily integrated in tools afterwards, either as CLI app or as a plugin. I'm so sick of seeing it again and again.

yeah, I thought on my first glance that it was an annotation processor somewhere we could reuse, or at least a utility program like spring boot or micronaut, I haven't found one though and following the gradle / maven plugins I ended up in this (I'm sure there will be more things that are blind spots down the line, because these plugins look massive, but hopefully we can give a basic web server / native support + testing to start with)

**Comment by vaslabs** [comment]
starting the work [here](https://github.com/com-lihaoyi/mill/pull/6771)

**Comment by vaslabs** [comment]
I think with the latest PR everything is in order. Once it's merged @souvlakias let's give it a go for the todo-mvc example , with the difference (compared to others) of having the quarkus native tests as well

can use the guide [here](https://quarkus.io/guides/web)

**Comment by vaslabs** [comment]
I think this can be closed now
